receiver/sapmreceiver/trace_receiver.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sapmreceiver
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/mux"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/observability"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	jaegertranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace/jaeger"
+	splunksapm "github.com/signalfx/sapm-proto/gen"
+	"github.com/signalfx/sapm-proto/sapmhandler"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+)
+
+const traceSource string = "sapm"
+
+// Receiver type is used to receive spans in the Splunk SAPM format over HTTP
+type sapmReceiver struct {
+	// mu protects the fields of this type
+	mu        sync.Mutex
+	startOnce sync.Once
+	stopOnce  sync.Once
+	logger    *zap.Logger
+
+	config     *Config
+	defaultCtx context.Context
+	server     *http.Server
+
+	nextConsumer consumer.TraceConsumer
+}
+
+// HTTPHandlerFunction returns an http.HandlerFunc that handles SAPM requests
+func (sr *sapmReceiver) HTTPHandlerFunc() func(rw http.ResponseWriter, req *http.Request) {
+	// receiverFn is a discrete function passed to sapm-proto's http.HanlderFunc.  This receiverFn actually processes the
+	// sapm and passes it on to the next trace consumer
+	receiverFn := func(ctx context.Context, sapm *splunksapm.PostSpansRequest, err error) error {
+		rCtx := observability.ContextWithReceiverName(ctx, "sapm")
+
+		// process the sapm
+		if sapm != nil {
+			for _, batch := range sapm.Batches {
+				// convert the jager batches to OCProto
+				td, err := jaegertranslator.ProtoBatchToOCProto(*batch)
+				if err != nil {
+					return err
+				}
+				td.SourceFormat = "sapm"
+
+				// pass the trace data to the next consumer
+				sr.nextConsumer.ConsumeTraceData(rCtx, td)
```

Comment:
I think `ConsumeTraceData` can return an error, it needs to be handled here otherwise we may have silent drops.

Reply:
@tigrannajaryan Is returning on error appropriate for converting the batches and consuming their trace data?  Or should I attempt to process the rest of the batches before returning?

Reply:
We have 2 options:
1. Return ErrBadRequest error to the caller. This will ensure caller does not re-send this anymore.
2. Return "Success" but increment local "bad input data" counter and process the rest of batches which are good.

(1) is definitely more appropriate if ProtoBatchToOCProto returns an error because it typically means something is wrong with span data.
(2) _may_ be more appropriate for `ConsumeTraceData` errors, but I could argue either way.

To keep it simple I would handle errors from both functions the same way and would just return ErrBadRequest (but make sure we actually return this exact error so that SAPM responds with 400 Bad Data and we don't receive this data again).